testrunner/src/mill/testrunner/TestRunnerUtils.scala

```diff
@@ -213,6 +228,111 @@ import scala.jdk.CollectionConverters.IteratorHasAsScala
     (doneMessage, results.toSeq)
   }
 
+  private def stealTaskFromSelectorFolder(
+      testClasses: Seq[(Class[?], Fingerprint)],
+      runner: Runner,
+      stealFolder: os.Path,
+      selectorFolder: os.Path
+  ): Array[Task] = {
+    while (true) {
+      val files = os.list(selectorFolder)
+      if (files.nonEmpty) {
+        val offset = Random.nextInt(files.size)
+        val stole =
+          try {
+            os.move(
+              files(offset),
+              stealFolder / s"selector-stolen",
+              atomicMove = true
+            )
+            true
+          } catch {
+            case e: Exception => false
+          }
+        if (stole) {
+          val selectors = os.read.lines(stealFolder / s"selector-stolen")
+          val classFilter = TestRunnerUtils.globFilter(selectors)
+          val tasks = runner.tasks(
+            for ((cls, fingerprint) <- testClasses.iterator.toArray if classFilter(cls.getName))
+              yield new TaskDef(
+                cls.getName.stripSuffix("$"),
+                fingerprint,
+                false,
+                Array(new SuiteSelector)
+              )
+          )
+          os.remove(stealFolder / s"selector-stolen")
+          return tasks
+        }
+      } else {
+        return Array.empty[Task]
+      }
+    }
+
+    // Compiler doesn't know that we returned from above, add this as a sanity check and assertion
+    ???
+  }
+
+  def stealTasks(
+      testClasses: Seq[(Class[?], Fingerprint)],
+      testReporter: TestReporter,
+      runner: Runner,
+      stealFolder: os.Path,
+      selectorFolder: os.Path
+  ): (String, Iterator[TestResult]) = {
+    // Capture this value outside of the task event handler so it
+    // isn't affected by a test framework's stream redirects
+    val systemOut = System.out
+    val events = new ConcurrentLinkedQueue[Event]()
+    val doneMessage = {
+
+      while ({
+        val tasks = stealTaskFromSelectorFolder(testClasses, runner, stealFolder, selectorFolder)
+        if (tasks.nonEmpty) {
+          executeTasks(tasks, testReporter, runner, events, systemOut)
+          true
+        } else {
+          false
+        }
+      }) ()
+
+      runner.done()
+    }
+
+    if (doneMessage != null && doneMessage.nonEmpty) {
+      if (doneMessage.endsWith("\n"))
+        println(doneMessage.stripSuffix("\n"))
+      else
+        println(doneMessage)
+    }
+
+    val results = parseRunTaskResults(events.iterator().asScala)
+
+    (doneMessage, results)
```

Comment:
There's a bunch of duplication here between `runTasks` and `stealTasks`; let's consolidate it into a helper method or something to keep things DRY